exchanges/poloniex/poloniex.go

```diff
@@ -3,933 +3,1124 @@ package poloniex
-		headers["Key"] = creds.Key
-		values.Set("nonce", e.Requester.GetNonce(nonce.UnixNano).String())
-		values.Set("command", endpoint)
-
-		hmac, err := crypto.GetHMAC(crypto.HashSHA512, []byte(values.Encode()), []byte(creds.Secret))
+		headers["Content-Type"] = "application/json"
+		headers["key"] = creds.Key
+		if values == nil {
+			values = url.Values{}
+		}
+		timestamp := time.Now()
+		bodyPayload := []byte("{}")
+		signTimestamp := strconv.FormatInt(timestamp.UnixMilli(), 10)
+		values.Set("signTimestamp", signTimestamp)
+		var signatureStrings string
+		switch method {
+		case http.MethodGet, "get":
+			signatureStrings = http.MethodGet + "\n" + path + "\n" + values.Encode()
+		default:
+			if body != nil {
+				bodyPayload, err = json.Marshal(body)
+				if err != nil {
+					return nil, err
+				}
+			}
+			if string(bodyPayload) != "{}" {
+				signatureStrings = method + "\n" + path + "\n" + "requestBody=" + string(bodyPayload) + "&" + values.Encode()
+			} else {
+				signatureStrings = method + "\n" + path + "\n" + values.Encode()
+			}
+		}
+		var hmac []byte
+		hmac, err = crypto.GetHMAC(crypto.HashSHA256,
+			[]byte(signatureStrings),
+			[]byte(creds.Secret))
 		if err != nil {
 			return nil, err
 		}
-		headers["Sign"] = hex.EncodeToString(hmac)
 
-		path := fmt.Sprintf("%s/%s", ePoint, poloniexAPITradingEndpoint)
-		return &request.Item{
-			Method:                 method,
-			Path:                   path,
-			Headers:                headers,
-			Body:                   bytes.NewBufferString(values.Encode()),
-			Result:                 result,
-			NonceEnabled:           true,
-			Verbose:                e.Verbose,
-			HTTPDebugging:          e.HTTPDebugging,
-			HTTPRecording:          e.HTTPRecording,
-			HTTPMockDataSliceLimit: e.HTTPMockDataSliceLimit,
-		}, nil
-	}, request.AuthenticatedRequest)
+		headers["signatureMethod"] = "hmacSHA256"
+		headers["signature"] = base64.StdEncoding.EncodeToString(hmac)
+		headers["signTimestamp"] = signTimestamp
+		values.Del("signTimestamp") // The signature timestamp has been removed from the query string as it is now included in the request header.
+
+		req := &request.Item{
+			Method:        method,
+			Path:          common.EncodeURLValues(endpoint+path, values),
+			Result:        resp,
+			Headers:       headers,
+			Verbose:       e.Verbose,
+			HTTPDebugging: e.HTTPDebugging,
+			HTTPRecording: e.HTTPRecording,
+		}
+		if method != http.MethodGet && len(bodyPayload) > 0 && string(bodyPayload) != "{}" {
+			req.Body = bytes.NewBuffer(bodyPayload)
+		}
+		return req, nil
+	}
+	if err := e.SendPayload(ctx, epl, requestFunc, request.AuthenticatedRequest); err != nil {
+		return fmt.Errorf("%w %w", request.ErrAuthRequestFailed, err)
+	}
+	if resp == nil {
+		return common.ErrNoResponse
+	}
+	if errType, ok := resp.(interface{ Error() error }); ok && errType.Error() != nil {
+		return fmt.Errorf("%w: %w", request.ErrAuthRequestFailed, errType.Error())
+	}
```

Comment:
🚧 Change request:
As above
🚧 Change request:
Same request as before.
* ~Please check `result` not `resp` for `nil`~
* Please check `resp` for the Error, not `result`
* **And actually, I think we can simplify things. The Error functions can do the nil check as appropriately for the type!**
```suggestion
	if errType, ok := resp.(interface{ Error() error }); ok && errType.Error() != nil {
		return fmt.Errorf("%w: %w", request.ErrAuthRequestFailed, errType.Error())
	}
```